dig.go

```diff
@@ -772,3 +785,122 @@ func shuffledCopy(rand *rand.Rand, items []reflect.Value) []reflect.Value {
 	}
 	return newItems
 }
+
+// EraseValueProvider provides a hack to erase the constructor and object of specified return type and name
+// The object depends on the removed object will be removed recursively in the next Invoke() call
+// Only one of name or group should be set.
+func (c *Container) EraseValueProvider(t reflect.Type, name, group string) {
+	k := key{
+		t:     t,
+		name:  name,
+		group: group,
+	}
+
+	// check if such provider exists
+	nodes, ok := c.providers[k]
+	if !ok {
+		// it is a shortcut
+		return
+	}
+
+	for _, n := range nodes {
+		if n.called == true {
+			// the constructor is called, thus we have to remove the corresponding value
+			if group != `` {
+				delete(c.groups, k)
+			} else {
+				delete(c.values, k)
+			}
+		}
+	}
+
+	// finally clear the providers and also rebuild the c.nodes
+	c.isVerifiedValid = false
+	delete(c.providers, k)
+	c.nodes = []*node{}
+	for _, array := range c.providers {
+		c.nodes = append(c.nodes, array...)
+	}
+}
+
+func (c *Container) eraseInvalidValues() {
+	// marks all called nodes as invalid first
+	for _, n := range c.nodes {
+		n.valid = false
+	}
+
+	// try to mark the node as valid, or remove the corresponding value
+	for k := range c.providers {
+		verifyNode(c, k)
+	}
+
+	// finally mark the verifyValid
+	c.isVerifiedValid = true
+}
+
+func getKeysFromParamList(c *Container, pl paramList) []key {
+	result := []key{}
+
+	fn := func(param param) bool {
+		switch p := param.(type) {
+		case paramSingle:
+			k := key{name: p.Name, t: p.Type}
+			result = append(result, k)
+			return false
+		case paramGroupedSlice:
+			// NOTE: The key uses the element type, not the slice type.
+			k := key{group: p.Group, t: p.Type.Elem()}
+			result = append(result, k)
+			return false
+		default:
+			// Recurse for non-edge params.
+			return true
+		}
+	}
+
+	walkParam(pl, paramVisitorFunc(fn))
+
+	return result
+}
+
+func verifyNode(c *Container, k key) (needClear bool) {
```

Comment:
`func verifyNode(c *Container, k key) bool {`

Reply:
it is okay as the output variable can hint the purpose for this output